File: TypeScript/src/P12IntegerToRoman.ts
const intToRoman = (num: number): string => {
  const valueSymbols: [number, string][] = [
    [1000, "M"],
    [900, "CM"],
    [500, "D"],
    [400, "CD"],
    [100, "C"],
    [90, "XC"],
    [50, "L"],
    [40, "XL"],
    [10, "X"],
    [9, "IX"],
    [5, "V"],
    [4, "IV"],
    [1, "I"],
  ];

  let roman: string = "";
  for (const [value, symbol] of valueSymbols) {
    while (num >= value) {
      roman += symbol;
      num -= value;
    }
  }

  return roman;
}

console.log("Problem 12:");
console.log(intToRoman(3));   // "III"
console.log(intToRoman(4));   // "IV"
console.log(intToRoman(9));   // "IX"
console.log(intToRoman(58));  // "LVIII"
console.log(intToRoman(1994)); // "MCMXCIV"

File: Java/src/main/java/org/adler/P12IntegerToRoman.java
package org.adler;

public class P12IntegerToRoman {
    public static String intToRoman(int num) {
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] symbols = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        StringBuilder roman = new StringBuilder();

        for (int i = 0; i < values.length; i++) {
            while (num >= values[i]) {
                roman.append(symbols[i]);
                num -= values[i];
            }
        }

        return roman.toString();
    }

    public static void main(String[] args) {
        System.out.println("Problem 12:");
        System.out.println(intToRoman(58));   // "LVIII"
        System.out.println(intToRoman(3));    // "III"
        System.out.println(intToRoman(4));    // "IV"
        System.out.println(intToRoman(9));    // "IX"

        System.out.println(intToRoman(1994)); // "MCMXCIV"
    }
}



